indexing-service/src/main/java/org/apache/druid/indexing/overlord/ThreadingTaskRunner.java

```diff
@@ -0,0 +1,710 @@
+    if (gracefulShutdownMillis > 0) {
+      try {
+        final boolean terminated = controlThreadExecutor.awaitTermination(
+            gracefulShutdownMillis,
+            TimeUnit.MILLISECONDS
+        );
+        final long elapsed = System.currentTimeMillis() - start.getMillis();
+        if (terminated) {
+          log.info("Finished stopping in %,dms.", elapsed);
+        } else {
+          final Set<String> stillRunning;
+          synchronized (tasks) {
+            stillRunning = ImmutableSet.copyOf(tasks.keySet());
+          }
+          log.makeAlert("Failed to stop task threads")
+             .addData("stillRunning", stillRunning)
+             .addData("elapsed", elapsed)
+             .emit();
+
+          log.warn(
+              "Executor failed to stop after %,dms, not waiting for it! Tasks still running: [%s]",
+              elapsed,
+              Joiner.on("; ").join(stillRunning)
+          );
+        }
+      }
+      catch (InterruptedException e) {
+        log.warn(e, "Interrupted while waiting for executor to finish.");
+        Thread.currentThread().interrupt();
+      }
+    } else {
+      log.warn("Ran out of time, not waiting for executor to finish!");
+    }
+  }
+
+  @Override
+  public Collection<? extends TaskRunnerWorkItem> getRunningTasks()
+  {
+    synchronized (tasks) {
+      final List<TaskRunnerWorkItem> ret = new ArrayList<>();
+      for (final ThreadingTaskRunnerWorkItem taskWorkItem : tasks.values()) {
+        if (taskWorkItem.getThread() != null) {
+          ret.add(taskWorkItem);
+        }
+      }
+      return ret;
+    }
+  }
+
+  @Override
+  public Collection<? extends TaskRunnerWorkItem> getPendingTasks()
+  {
+    synchronized (tasks) {
+      final List<TaskRunnerWorkItem> ret = new ArrayList<>();
+      for (final ThreadingTaskRunnerWorkItem taskWorkItem : tasks.values()) {
+        if (taskWorkItem.getThread() == null) {
+          ret.add(taskWorkItem);
+        }
+      }
+      return ret;
+    }
+  }
+
+  @Override
+  public Collection<? extends TaskRunnerWorkItem> getKnownTasks()
+  {
+    synchronized (tasks) {
+      return Lists.newArrayList(tasks.values());
+    }
+  }
+
+  @Nullable
+  @Override
+  public RunnerTaskState getRunnerTaskState(String taskId)
+  {
+    final ThreadingTaskRunnerWorkItem workItem = tasks.get(taskId);
+    if (workItem == null) {
+      return null;
+    } else {
+      if (workItem.getThread() == null) {
+        return RunnerTaskState.PENDING;
+      } else if (!workItem.isFinished()) {
+        return RunnerTaskState.RUNNING;
+      } else {
+        return RunnerTaskState.NONE;
+      }
+    }
+  }
+
+  @Override
+  public Optional<ScalingStats> getScalingStats()
+  {
+    return Optional.absent();
+  }
+
+
+  // Save running tasks to a file, so they can potentially be restored on next startup. Suppresses exceptions that
+  // occur while saving.
+  @GuardedBy("tasks")
+  private void saveRunningTasks()
+  {
```

Comment:
There seems to be a fair bit of overlap here and with `ForkingTaskRunner`, same `TaskRestoreInfo`, same restore methods, same get task lists methods; could you look into maybe extracting a base type of `TaskRunner` so we can avoid duplicating this stuff, and then these implementations can just provide `TaskRunnerWorkItem` implementations and the logic to run and manage them.

Reply:
Extracted a base `BaseRestorableTaskRunner` class